### PR TITLE
Add a validation for rancher-nfs and the driver will stop when validation fail

### PR DESCRIPTION
**Pull request by orangedeng**
Add a validation for nfs.
Fixed:
https://github.com/rancher/rancher/issues/9608
https://github.com/rancher/rancher/issues/9607

@yasker

### CODE REVIEW

File: package/nfs/rancher-nfs [hunk 1]
```diff
@@ -182,5 +183,24 @@ delete() {
...
 
+validate() {
+
+    # default configuration
+    local host="$NFS_SERVER"
+    local exportDir="$MOUNT_DIR"
+    local opts="$MOUNT_OPTS"
+    local mountDir="$(tmp_dir)"
+
+    log_info "validation" "Mounting" "$mountDir" "for validating NFS configuration."
```
Comment (yasker) [inline package/nfs/rancher-nfs]: Those separate quotes are unnecessary.

"Validating NFS by mounting to $mountDir" should be enough.

File: package/nfs/rancher-nfs [hunk 2]
```diff
@@ -182,5 +183,24 @@ delete() {
     print_success purged
 }
 
+validate() {
+
```
Comment (yasker) [inline package/nfs/rancher-nfs]: Unnecessary empty line.

**Review by yasker — CHANGES_REQUESTED** [review]
Also, how long it will take it to fail?

**Comment by orangedeng** [comment]
It will take about 2 mins when there is a connection problem(such as wrong ip address or wrong port). But it will fail immediately when there are wrong nfs parameters.

File: package/nfs/rancher-nfs (hunk 1 above)
Comment (orangedeng) [reply package/nfs/rancher-nfs]: The first parameter of function `log_info` needs a name, so I put a `validation` into it. I will make it right.

**Review by yasker — APPROVED** [review]

**Comment by yasker** [comment]
Next time remember to put the bug URL in the commit message.

**Comment by yasker** [comment]
Please make a PR to `rancher-nfs` catalog for `rancher-catalog` on `v1.6-development` branch for `rancher-nfs v0.8.9`

**Comment by kingsd041** [comment]
I build a new image and use it, when I input the wrong NFS parameters, NFS container  will restart 2 minutes at a time, output log: ```time = "of" the 2017-09-27 T09:24:40 Z level = fatal MSG = "Failed to initialize: Failed mount -o proto = TCP, nfsvers = 4 192.168.99.206: / opt/nfsdata pvq2"/TMP / 1``` 

State is active.
I don't think the NFS container state is active when the parameters are wrong. So the user doesn't know where the problem is.

**Comment by yasker** [comment]
@kingsd041 You meant, you don't think the NFS container state "should be" active when parameters are wrong, right?

**Comment by kingsd041** [comment]
@yasker  yes

**Comment by orangedeng** [comment]
The wrong IP address is not the problem I can detect. Because the NFS client needs to wait for the connection time out. The container will be active while it is connecting.

**Comment by yasker** [comment]
@kingsd041 has a valid point. What user from the UI is:
1. Service started up
2. Service is Green.
3. After 2 minutes, service reboots.
4. Back to step 2.

From user's perspective, it's strange that a service kept rebooting, but doesn't show any error state.

I am thinking to add health check to rancher-storage. But we need to set health check    
 initializing_timeout and reinitializing_timeout to more than 120 seconds.

@cjellick what do you think?

**Comment by cjellick** [comment]
Stopping storage driver containers is dangerous: if a volume has ever been created using the driver, then the docker daemon will freeze on `docker ps` if the storage driver container is stoppped.
@yasker I like your idea about adding a health check, but can we make it an information-only healthcheck (meaning the `strategy: none`).

Here's what I suggest:
Do not cause the storage container to exit if validation fails, instead just cause an information-only healthcheck to fail. You could do this by adding a validate function to the framework that drivers may implement. Then, the go code could call it and store the result in menory and use the result for responding to the healtcheck. As stated, the initialization/reinitialization period would need to be greater than the 2 minute timeout.

The logs for the container should as clearly as possible state whats going on sot that if a user looks at container that is still intiializing it is clear why it is doing so and if it is unhealthy, the logs should clearly indicate why.

Finally, we could further enhance this by having the go framework check the validation state of the driver. If it receives non-read action like create, mount, or unmount, it could immediately fail if the validation state stored in memory says the driver is not healthy.

**Pull request merged**